cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/PreviewRunnerTwillRunnable.java

```diff
@@ -238,36 +243,46 @@ protected void configure() {
+    // metadata stored in AppFabric.
     modules.add(new DFSLocationModule());
-    modules.add(new MetadataServiceModule());
-    modules.add(new CoreSecurityRuntimeModule().getInMemoryModules());
+    // Configurator tasks should be executed in-memory since it is in the preview runner pod.
+    modules.add(new FactoryModuleBuilder().implement(Configurator.class, InMemoryConfigurator.class)
+                  .build(ConfiguratorFactory.class));
+
     modules.add(new AuthenticationContextModules().getMasterWorkerModule());
-    modules.add(new AuthorizationModule());
     modules.add(new AuthorizationEnforcementModule().getNoOpModules());
-    modules.add(Modules.override(
-      new DataFabricModules("master").getDistributedModules()).with(new AbstractModule() {
+    modules.add(new AbstractModule() {
       @Override
       protected void configure() {
-        // Bind transaction system to a constant one, basically no transaction, with every write become
-        // visible immediately.
-        // TODO: Ideally we shouldn't need this at all. However, it is needed now to satisfy dependencies
-        bind(TransactionSystemClientService.class).to(DelegatingTransactionSystemClientService.class);
         bind(TransactionSystemClient.class).to(ConstantTransactionSystemClient.class);
 
-        bind(ExploreClient.class).to(UnsupportedExploreClient.class);
         bind(PreviewRequestPollerInfoProvider.class).toInstance(() -> pollerInfoBytes);
-      }
-    }));
 
+        // Artifact Repository should use RemoteArtifactRepository.
+        // TODO: Consider adding a remote artifact respository handler to the preview manager so that preview runners
+        // do not have to talk directly to app-fabric for artifacts. This will prevent hot-spotting app-fabric.
+        bind(ArtifactRepositoryReader.class).to(RemoteArtifactRepositoryReader.class).in(Scopes.SINGLETON);
+        bind(ArtifactRepository.class).to(RemoteArtifactRepository.class);
+        // Use artifact localizer client if it is enabled, otherwise use regular RemotePluginFinder
+        if (cConf.getBoolean(Constants.Preview.ARTIFACT_LOCALIZER_ENABLED)) {
+          bind(PluginFinder.class).to(RemoteWorkerPluginFinder.class);
+          bind(ArtifactLocalizerClient.class).in(Scopes.SINGLETON);
+        } else {
+          bind(PluginFinder.class).to(RemotePluginFinder.class);
+        }
+        // Preview runner pods should not have any elevated privileges, so use the current UGI.
+        bind(UGIProvider.class).to(CurrentUGIProvider.class);
+        // Need ProgramRunnerFactory for the RemoteArtifactRepository
+        bind(ProgramRunnerFactory.class).to(DefaultProgramRunnerFactory.class).in(Scopes.SINGLETON);
+        MapBinder.newMapBinder(binder(), ProgramType.class, ProgramRunner.class);
+        bind(ProgramStateWriter.class).to(MessagingProgramStateWriter.class);
+        bind(ProgramRuntimeProvider.Mode.class).toInstance(ProgramRuntimeProvider.Mode.LOCAL);
```

Comment:
Are these binding required because you removed `AppFabricServiceRuntimeModule` ?

Reply:
Yup